watch/watch_test.go

```diff
@@ -1,15 +1,87 @@
 package watch
 
 import (
+	"context"
 	"sync"
 	"testing"
+	"time"
 
 	"github.com/docker/go-events"
+	"github.com/stretchr/testify/require"
 )
 
+func TestTimeoutLimitWatch(t *testing.T) {
+	require := require.New(t)
+	q := NewQueue(WithTimeout(time.Second), WithLimit(5), WithCloseOutChan())
+	defer q.Close()
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// Cancelling a watcher's context should remove the watcher from the queue and
+	// close its channel.
+	doneChan := make(chan struct{})
+	go func() {
+		events := q.WatchContext(ctx)
+		for range events {
+		}
+		close(doneChan)
+	}()
+	cancel()
+	<-doneChan
+
+	// Test a scenario with a faster write rate than read rate The queue
+	// should eventually fill up and the channel will be closed.
+	readerSleepDuration := 100 * time.Millisecond
+	writerSleepDuration := 10 * time.Millisecond
+
+	events, cancel := q.Watch()
+	defer cancel()
+
+	receivedChan := make(chan struct{})
+	eventsClosed := false
+
+	go func() {
+		closed := false
+		for range events {
+			// After receiving the first event, block indefinitely
+			if !closed {
+				close(receivedChan)
+				closed = true
+			}
+			time.Sleep(readerSleepDuration)
+		}
+		eventsClosed = true
+	}()
+
+	// Publish one event and wait for the watcher to receive it
+	q.Publish("new event")
+	<-receivedChan
+
+	doneChan = make(chan struct{})
+	go func() {
+		for !eventsClosed {
```

Comment:
I think the race detector will consider this a data race. If we make it a channel that gets closed, it avoids the problem. I'd kind of like to unify this goroutine with the `select` below anyway. How about something like this:

```go
timeoutTimer := time.NewTimer(time.Minute)
defer timeoutTimer.Stop()

selectLoop:
for {
        select {
        case <-eventsClosed:
                break selectLoop
        case <-time.After(writerSleepDuration):
                q.Publish("new event")
        case <-timeoutTimer.C:
                require.Fail("Timeout exceeded")
        }
}
```